R/dist_spec.R

```diff
@@ -34,151 +22,67 @@
+#' @importFrom rlang arg_match
+discrete_pmf <- function(distribution =
+                           c("exp", "gamma", "lognormal", "normal", "fixed"),
+                         params, max_value, tolerance, width) {
+  distribution <- arg_match(distribution)
+  ## define unnormalised support function and cumulative density function
+  if (distribution == "exp") {
     updist <- function(n) {
       pexp(n, params[["rate"]])
     }
-  } else if (model == "gamma") {
+    qdist <- qexp
+  } else if (distribution == "gamma") {
     updist <- function(n) {
       pgamma(n, params[["shape"]], params[["rate"]])
     }
-  } else if (model == "lognormal") {
+    qdist <- qgamma
+  } else if (distribution == "lognormal") {
     updist <- function(n) {
       plnorm(n, params[["meanlog"]], params[["sdlog"]])
     }
-  } else if (model == "normal") {
+    qdist <- qlnorm
+  } else if (distribution == "normal") {
     updist <- function(n) {
       pnorm(n, params[["mean"]], params[["sd"]])
     }
-  } else if (model == "fixed") {
+    qdist <- qnorm
+  } else if (distribution == "fixed") {
     updist <- function(n) {
       as.integer(n > params[["value"]])
     }
+    qdist <- function(p, value) return(value)
   }
 
-  if (discrete) {
-    cmf <- c(0, updist(1),
-      updist(seq_len(max_value)) + updist(seq_len(max_value) + 1)
-    ) /
-      (updist(max_value) + updist(max_value + 1))
-    pmf <- diff(cmf)
-    rdist <- function(n) {
-      sample(
-        x = seq_len(max_value + 1) - 1, size = n, prob = pmf, replace = TRUE
-      )
-    }
-    pdist <- function(n) {
-      cmf[n + 1]
-    }
-    ddist <- function(n) {
-      pmf[n + 1]
-    }
-  } else {
-    pdist <- function(n) {
-      updist(n) / updist(max_value + 1)
-    }
-    ddist <- function(n) {
-      pdist(n + 1) - pdist(n)
-    }
-    if (model == "exp") {
-      rdist <- function(n) {
-        rexp(n, params[["rate"]])
-      }
-    } else if (model == "gamma") {
-      rdist <- function(n) {
-        rgamma(n, params[["shape"]], params[["rate"]])
-      }
-    } else if (model == "lognormal") {
-      rdist <- function(n) {
-        rlnorm(n, params[["meanlog"]], params[["sdlog"]])
-      }
+  ## apply tolerance if given
+  if (!missing(tolerance)) {
+    ## tolerance_max
+    tol_max <- do.call(qdist, c(list(p = 1 - tolerance), params))
+    if (missing(max_value) || tol_max < max_value) {
+      max_value <- tol_max
     }
   }
 
-  # define internal sampling function
-  inner_skel <- function(n, dist = FALSE, cum = TRUE, max_value = NULL) {
-    if (dist) {
-      if (cum) {
-        ret <- pdist(n)
-      } else {
-        ret <- ddist(n)
-      }
-      ret[ret > 1] <- NA_real_
-      return(ret)
-    } else {
-      rdist(n)
-    }
+  ## determine pmf
+  max_value <- ceiling(max_value)
+  if (max_value < width) {
+    cmf <- c(0, 1)
+  } else {
+    x <- seq(width, max_value, by = width)
+    cmf <- c(0, updist(width), (updist(x) + updist(x + width))) /
```

Comment:
docs should probably mention what kind of discretisation this is and if its an approximation (which I think it is?)

Reply:
you mean adding to this?
https://github.com/epiforecasts/EpiNow2/blob/e95da199cfd77b79daca81e369fdbef33bc67bc9/R/dist_spec.R#L8

or somewhere else?

Reply:
Addressed in 0a0538ca523a1f403a534a0e4f49a1ca69aa145f